File: src/core/StackChecker.ts
import type { PlacedCargo, CargoItemDef, Vec3 } from './types'
import { computeRotatedAABB } from './Voxelizer'

export interface StackViolation {
  instanceId: number
  cargoDefId: string
  name: string
  maxStackWeightKg: number
  actualStackWeightKg: number
}

interface PlacedAABB {
  instanceId: number
  defId: string
  min: Vec3
  max: Vec3
  weightKg: number
}

const EPSILON = 1.5 // tolerance in cm for "touching"

/**
 * Binary search: find the leftmost index where sorted[index].minY >= target.
 * sorted is an array of { origIdx, minY } sorted by minY ascending.
 */
function lowerBound(sorted: { origIdx: number; minY: number }[], target: number): number {
  let lo = 0
  let hi = sorted.length
  while (lo < hi) {
    const mid = (lo + hi) >>> 1
    if (sorted[mid]!.minY < target) lo = mid + 1
    else hi = mid
  }
  return lo
}

/**
 * Build the support graph using Y-sort + binary search.
 * For each item a, find items b where b.min.y ≈ a.max.y (b sits on a),
 * then check XZ overlap only for those candidates.
 */
function buildSupportGraph(
  aabbs: PlacedAABB[],
): { onTopOf: number[][]; supportedBy: number[][] } {
  const n = aabbs.length
  const onTopOf: number[][] = Array.from({ length: n }, () => [])
  const supportedBy: number[][] = Array.from({ length: n }, () => [])

  // Sort indices by min.y for binary search
  const sorted = aabbs.map((a, i) => ({ origIdx: i, minY: a.min.y }))
  sorted.sort((a, b) => a.minY - b.minY)

  // For each item a, find candidates b where b.min.y ∈ [a.max.y - EPSILON, a.max.y + EPSILON)
  for (let i = 0; i < n; i++) {
    const a = aabbs[i]!
    const lo = lowerBound(sorted, a.max.y - EPSILON)

    for (let k = lo; k < sorted.length; k++) {
      const entry = sorted[k]!
      if (entry.minY >= a.max.y + EPSILON) break
      const j = entry.origIdx
      if (j === i) continue

      const b = aabbs[j]!
      // Check XZ overlap
      const overlapX = a.min.x < b.max.x && a.max.x > b.min.x
      const overlapZ = a.min.z < b.max.z && a.max.z > b.min.z
      if (!overlapX || !overlapZ) continue

      // b sits on a (b.min.y ≈ a.max.y)
      onTopOf[i]!.push(j)
      supportedBy[j]!.push(i)
    }
  }

  return { onTopOf, supportedBy }
}

/**
 * Check stack constraints for all placements.
 * Returns violations where weight on top exceeds maxStackWeightKg or noStack is violated.
 */
export function checkStackConstraints(
  placements: PlacedCargo[],
  cargoDefs: CargoItemDef[],
): StackViolation[] {
  if (placements.length === 0) return []

  const defMap = new Map<string, CargoItemDef>()
  for (const d of cargoDefs) {
    defMap.set(d.id, d)
  }

  // Compute AABBs for all placements
  const aabbs: PlacedAABB[] = []
  for (const p of placements) {
    const def = defMap.get(p.cargoDefId)
    if (!def) continue

    const aabb = computeRotatedAABB(
      def.widthCm, def.heightCm, def.depthCm,
      p.positionCm, p.rotationDeg,
    )
    aabbs.push({
      instanceId: p.instanceId,
      defId: def.id,
      min: aabb.min,
      max: aabb.max,
      weightKg: def.weightKg,
    })
  }

  // Build support graph using Y-sort + binary search (O(n log n) instead of O(n²))
  const { onTopOf } = buildSupportGraph(aabbs)

  // For each constrained item, compute total weight above it recursively (with memoization)
  const weightAboveCache = new Map<number, number>()

  function getWeightAbove(idx: number): number {
    const cached = weightAboveCache.get(idx)
    if (cached !== undefined) return cached

    let total = 0
    for (const aboveIdx of onTopOf[idx]!) {
      const aboveItem = aabbs[aboveIdx]!
      total += aboveItem.weightKg + getWeightAbove(aboveIdx)
    }

    weightAboveCache.set(idx, total)
    return total
  }

  const violations: StackViolation[] = []

  for (let i = 0; i < aabbs.length; i++) {
    const item = aabbs[i]!
    const def = defMap.get(item.defId)
    if (!def) continue

    const maxStack = def.noStack ? 0 : def.maxStackWeightKg
    if (maxStack === undefined) continue

    const actualWeight = getWeightAbove(i)
    if (actualWeight > maxStack) {
      violations.push({
        instanceId: item.instanceId,
        cargoDefId: def.id,
        name: def.name,
        maxStackWeightKg: maxStack,
        actualStackWeightKg: actualWeight,
      })
    }
  }

  return violations
}

/**
 * Check stack constraints using pre-computed AABBs.
 * Same logic as checkStackConstraints but avoids recomputing AABBs.
 */
export function checkStackConstraintsWithAABBs(
  placements: PlacedCargo[],
  cargoDefs: CargoItemDef[],
  aabbs: Array<{ min: Vec3; max: Vec3 }>,
): StackViolation[] {
  if (placements.length === 0) return []

  const defMap = new Map<string, CargoItemDef>()
  for (const d of cargoDefs) {
    defMap.set(d.id, d)
  }

  // Build PlacedAABB array from pre-computed AABBs
  const placedAABBs: PlacedAABB[] = []
  for (let i = 0; i < placements.length; i++) {
    const p = placements[i]!
    const def = defMap.get(p.cargoDefId)
    if (!def) continue

    const aabb = aabbs[i]!
    placedAABBs.push({
      instanceId: p.instanceId,
      defId: def.id,
      min: aabb.min,
      max: aabb.max,
      weightKg: def.weightKg,
    })
  }

  // Build support graph using Y-sort + binary search
  const { onTopOf } = buildSupportGraph(placedAABBs)

  // For each constrained item, compute total weight above it recursively (with memoization)
  const weightAboveCache = new Map<number, number>()

  function getWeightAbove(idx: number): number {
    const cached = weightAboveCache.get(idx)
    if (cached !== undefined) return cached

    let total = 0
    for (const aboveIdx of onTopOf[idx]!) {
      const aboveItem = placedAABBs[aboveIdx]!
      total += aboveItem.weightKg + getWeightAbove(aboveIdx)
    }

    weightAboveCache.set(idx, total)
    return total
  }

  const violations: StackViolation[] = []

  for (let i = 0; i < placedAABBs.length; i++) {
    const item = placedAABBs[i]!
    const def = defMap.get(item.defId)
    if (!def) continue

    const maxStack = def.noStack ? 0 : def.maxStackWeightKg
    if (maxStack === undefined) continue

    const actualWeight = getWeightAbove(i)
    if (actualWeight > maxStack) {
      violations.push({
        instanceId: item.instanceId,
        cargoDefId: def.id,
        name: def.name,
        maxStackWeightKg: maxStack,
        actualStackWeightKg: actualWeight,
      })
    }
  }

  return violations
}

// --- Incremental stack checking for AutoPacker ---

export interface StackContext {
  aabbs: PlacedAABB[]
  onTopOf: number[][]
  supportedBy: number[][]
  weightAboveCache: Map<number, number>
  defMap: Map<string, CargoItemDef>
}

/** Build an incremental stack context from existing placements. */
export function buildStackContext(
  placements: PlacedCargo[],
  defs: CargoItemDef[],
): StackContext {
  const defMap = new Map<string, CargoItemDef>()
  for (const d of defs) defMap.set(d.id, d)

  const aabbs: PlacedAABB[] = []
  for (const p of placements) {
    const def = defMap.get(p.cargoDefId)
    if (!def) continue
    const aabb = computeRotatedAABB(def.widthCm, def.heightCm, def.depthCm, p.positionCm, p.rotationDeg)
    aabbs.push({ instanceId: p.instanceId, defId: def.id, min: aabb.min, max: aabb.max, weightKg: def.weightKg })
  }

  // Build support graph using Y-sort + binary search (O(n log n) instead of O(n²))
  const { onTopOf, supportedBy } = buildSupportGraph(aabbs)

  return { aabbs, onTopOf, supportedBy, weightAboveCache: new Map(), defMap }
}

/**
 * Check if adding newPlacement would violate any stack constraints.
 * Only checks items affected by the new placement (O(N) instead of O(N²)).
 */
export function checkStackIncremental(
  ctx: StackContext,
  newPlacement: PlacedCargo,
  newDef: CargoItemDef,
): StackViolation[] {
  const aabb = computeRotatedAABB(
    newDef.widthCm, newDef.heightCm, newDef.depthCm,
    newPlacement.positionCm, newPlacement.rotationDeg,
  )
  const newAABB: PlacedAABB = {
    instanceId: newPlacement.instanceId,
    defId: newDef.id,
    min: aabb.min,
    max: aabb.max,
    weightKg: newDef.weightKg,
  }

  // Find which existing items the new placement sits on top of
  const directSupporters: number[] = []
  for (let i = 0; i < ctx.aabbs.length; i++) {
    const existing = ctx.aabbs[i]!
    const overlapX = existing.min.x < newAABB.max.x && existing.max.x > newAABB.min.x
    const overlapZ = existing.min.z < newAABB.max.z && existing.max.z > newAABB.min.z
    if (!overlapX || !overlapZ) continue
    if (Math.abs(newAABB.min.y - existing.max.y) < EPSILON) {
      directSupporters.push(i)
    }
  }

  if (directSupporters.length === 0) return []

  // The new placement has nothing above it, so its weight contribution
  // to supporters is just newDef.weightKg.
  // Check all transitive supporters downward.
  const violations: StackViolation[] = []
  const visited = new Set<number>()
  const queue = [...directSupporters]

  // Temporarily add newDef.weightKg to the weight-above chain
  const addedWeight = newDef.weightKg

  while (queue.length > 0) {
    const idx = queue.pop()!
    if (visited.has(idx)) continue
    visited.add(idx)

    const item = ctx.aabbs[idx]!
    const def = ctx.defMap.get(item.defId)
    if (def) {
      const maxStack = def.noStack ? 0 : def.maxStackWeightKg
      if (maxStack !== undefined) {
        const currentWeightAbove = getWeightAboveFromCtx(ctx, idx)
        if (currentWeightAbove + addedWeight > maxStack) {
          violations.push({
            instanceId: item.instanceId,
            cargoDefId: def.id,
            name: def.name,
            maxStackWeightKg: maxStack,
            actualStackWeightKg: currentWeightAbove + addedWeight,
          })
        }
      }
    }

    // Propagate downward through supporters
    for (const belowIdx of ctx.supportedBy[idx]!) {
      if (!visited.has(belowIdx)) queue.push(belowIdx)
    }
  }

  return violations
}

/** Add a confirmed placement to the stack context. */
export function addToStackContext(
  ctx: StackContext,
  placement: PlacedCargo,
  def: CargoItemDef,
): void {
  const aabb = computeRotatedAABB(
    def.widthCm, def.heightCm, def.depthCm,
    placement.positionCm, placement.rotationDeg,
  )
  const newIdx = ctx.aabbs.length
  ctx.aabbs.push({
    instanceId: placement.instanceId,
    defId: def.id,
    min: aabb.min,
    max: aabb.max,
    weightKg: def.weightKg,
  })
  ctx.onTopOf.push([])
  ctx.supportedBy.push([])

  // Build edges for the new item
  for (let i = 0; i < newIdx; i++) {
    const existing = ctx.aabbs[i]!
    const overlapX = existing.min.x < aabb.max.x && existing.max.x > aabb.min.x
    const overlapZ = existing.min.z < aabb.max.z && existing.max.z > aabb.min.z
    if (!overlapX || !overlapZ) continue

    // New item sits on existing
    if (Math.abs(aabb.min.y - existing.max.y) < EPSILON) {
      ctx.onTopOf[i]!.push(newIdx)
      ctx.supportedBy[newIdx]!.push(i)
    }
    // Existing sits on new item
    if (Math.abs(existing.min.y - aabb.max.y) < EPSILON) {
      ctx.onTopOf[newIdx]!.push(i)
      ctx.supportedBy[i]!.push(newIdx)
    }
  }

  // Invalidate weight-above cache for affected items (transitive supporters)
  const toInvalidate = new Set<number>()
  const queue = [...ctx.supportedBy[newIdx]!]
  while (queue.length > 0) {
    const idx = queue.pop()!
    if (toInvalidate.has(idx)) continue
    toInvalidate.add(idx)
    for (const belowIdx of ctx.supportedBy[idx]!) {
      if (!toInvalidate.has(belowIdx)) queue.push(belowIdx)
    }
  }
  for (const idx of toInvalidate) {
    ctx.weightAboveCache.delete(idx)
  }
}

function getWeightAboveFromCtx(ctx: StackContext, idx: number): number {
  const cached = ctx.weightAboveCache.get(idx)
  if (cached !== undefined) return cached

  let total = 0
  for (const aboveIdx of ctx.onTopOf[idx]!) {
    const aboveItem = ctx.aabbs[aboveIdx]!
    total += aboveItem.weightKg + getWeightAboveFromCtx(ctx, aboveIdx)
  }

  ctx.weightAboveCache.set(idx, total)
  return total
}

File: rust/autopack/src/stack_checker.rs
use std::collections::{HashMap, HashSet};

use crate::types::{CargoItemDef, PlacedCargo, Vec3};
use crate::voxelizer::compute_rotated_aabb;

/// Tolerance in cm for "touching" detection.
const EPSILON: f64 = 1.5;

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct StackViolation {
    pub instance_id: u32,
    pub cargo_def_id: String,
    pub name: String,
    pub max_stack_weight_kg: f64,
    pub actual_stack_weight_kg: f64,
}

#[derive(Debug, Clone)]
struct PlacedAABB {
    instance_id: u32,
    def_id: String,
    min: Vec3,
    max: Vec3,
    weight_kg: f64,
}

pub struct StackContext {
    aabbs: Vec<PlacedAABB>,
    on_top_of: Vec<Vec<usize>>,
    supported_by: Vec<Vec<usize>>,
    weight_above_cache: HashMap<usize, f64>,
    def_map: HashMap<String, CargoItemDef>,
}

impl StackContext {
    fn get_weight_above(&mut self, idx: usize) -> f64 {
        if let Some(&cached) = self.weight_above_cache.get(&idx) {
            return cached;
        }

        let above_indices: Vec<usize> = self.on_top_of[idx].clone();
        let mut total = 0.0;
        for above_idx in above_indices {
            total += self.aabbs[above_idx].weight_kg + self.get_weight_above(above_idx);
        }

        self.weight_above_cache.insert(idx, total);
        total
    }
}

/// Build an incremental stack context from existing placements.
pub fn build_stack_context(placements: &[PlacedCargo], defs: &[CargoItemDef]) -> StackContext {
    let def_map: HashMap<String, CargoItemDef> =
        defs.iter().map(|d| (d.id.clone(), d.clone())).collect();

    let mut aabbs = Vec::new();
    for p in placements {
        let Some(def) = def_map.get(&p.cargo_def_id) else {
            continue;
        };
        let aabb = compute_rotated_aabb(
            def.width_cm,
            def.height_cm,
            def.depth_cm,
            p.position_cm,
            p.rotation_deg,
        );
        aabbs.push(PlacedAABB {
            instance_id: p.instance_id,
            def_id: def.id.clone(),
            min: aabb.min,
            max: aabb.max,
            weight_kg: def.weight_kg,
        });
    }

    let n = aabbs.len();
    let mut on_top_of: Vec<Vec<usize>> = (0..n).map(|_| Vec::new()).collect();
    let mut supported_by: Vec<Vec<usize>> = (0..n).map(|_| Vec::new()).collect();

    for i in 0..n {
        for j in (i + 1)..n {
            let a = &aabbs[i];
            let b = &aabbs[j];
            let overlap_x = a.min.x < b.max.x && a.max.x > b.min.x;
            let overlap_z = a.min.z < b.max.z && a.max.z > b.min.z;
            if !overlap_x || !overlap_z {
                continue;
            }
            if (b.min.y - a.max.y).abs() < EPSILON {
                on_top_of[i].push(j);
                supported_by[j].push(i);
            }
            if (a.min.y - b.max.y).abs() < EPSILON {
                on_top_of[j].push(i);
                supported_by[i].push(j);
            }
        }
    }

    StackContext {
        aabbs,
        on_top_of,
        supported_by,
        weight_above_cache: HashMap::new(),
        def_map,
    }
}

/// Check if adding new_placement would violate any stack constraints.
/// Only checks items affected by the new placement (O(N) instead of O(N²)).
pub fn check_stack_incremental(
    ctx: &mut StackContext,
    new_placement: &PlacedCargo,
    new_def: &CargoItemDef,
) -> Vec<StackViolation> {
    let aabb = compute_rotated_aabb(
        new_def.width_cm,
        new_def.height_cm,
        new_def.depth_cm,
        new_placement.position_cm,
        new_placement.rotation_deg,
    );
    let new_aabb = PlacedAABB {
        instance_id: new_placement.instance_id,
        def_id: new_def.id.clone(),
        min: aabb.min,
        max: aabb.max,
        weight_kg: new_def.weight_kg,
    };

    // Find which existing items the new placement sits on top of
    let mut direct_supporters = Vec::new();
    for i in 0..ctx.aabbs.len() {
        let existing = &ctx.aabbs[i];
        let overlap_x = existing.min.x < new_aabb.max.x && existing.max.x > new_aabb.min.x;
        let overlap_z = existing.min.z < new_aabb.max.z && existing.max.z > new_aabb.min.z;
        if !overlap_x || !overlap_z {
            continue;
        }
        if (new_aabb.min.y - existing.max.y).abs() < EPSILON {
            direct_supporters.push(i);
        }
    }

    if direct_supporters.is_empty() {
        return Vec::new();
    }

    let added_weight = new_def.weight_kg;
    let mut violations = Vec::new();
    let mut visited = HashSet::new();
    let mut queue: Vec<usize> = direct_supporters;

    while let Some(idx) = queue.pop() {
        if !visited.insert(idx) {
            continue;
        }

        // Extract needed data before mutable borrow
        let item_instance_id = ctx.aabbs[idx].instance_id;
        let item_def_id = ctx.aabbs[idx].def_id.clone();

        // Clone constraint info to avoid borrow conflict with get_weight_above
        let constraint_info = ctx.def_map.get(&item_def_id).map(|def| {
            let max_stack = if def.no_stack == Some(true) {
                Some(0.0)
            } else {
                def.max_stack_weight_kg
            };
            (def.id.clone(), def.name.clone(), max_stack)
        });

        if let Some((def_id, def_name, Some(max_stack))) = constraint_info {
            let current_weight_above = ctx.get_weight_above(idx);
            if current_weight_above + added_weight > max_stack {
                violations.push(StackViolation {
                    instance_id: item_instance_id,
                    cargo_def_id: def_id,
                    name: def_name,
                    max_stack_weight_kg: max_stack,
                    actual_stack_weight_kg: current_weight_above + added_weight,
                });
            }
        }

        // Propagate downward through supporters
        for &below_idx in &ctx.supported_by[idx] {
            if !visited.contains(&below_idx) {
                queue.push(below_idx);
            }
        }
    }

    violations
}

/// Add a confirmed placement to the stack context.
pub fn add_to_stack_context(
    ctx: &mut StackContext,
    placement: &PlacedCargo,
    def: &CargoItemDef,
) {
    let aabb = compute_rotated_aabb(
        def.width_cm,
        def.height_cm,
        def.depth_cm,
        placement.position_cm,
        placement.rotation_deg,
    );
    let new_idx = ctx.aabbs.len();
    ctx.aabbs.push(PlacedAABB {
        instance_id: placement.instance_id,
        def_id: def.id.clone(),
        min: aabb.min,
        max: aabb.max,
        weight_kg: def.weight_kg,
    });
    ctx.on_top_of.push(Vec::new());
    ctx.supported_by.push(Vec::new());

    // Build edges for the new item
    for i in 0..new_idx {
        let existing = &ctx.aabbs[i];
        let overlap_x = existing.min.x < aabb.max.x && existing.max.x > aabb.min.x;
        let overlap_z = existing.min.z < aabb.max.z && existing.max.z > aabb.min.z;
        if !overlap_x || !overlap_z {
            continue;
        }

        // New item sits on existing
        if (aabb.min.y - existing.max.y).abs() < EPSILON {
            ctx.on_top_of[i].push(new_idx);
            ctx.supported_by[new_idx].push(i);
        }
        // Existing sits on new item
        if (existing.min.y - aabb.max.y).abs() < EPSILON {
            ctx.on_top_of[new_idx].push(i);
            ctx.supported_by[i].push(new_idx);
        }
    }

    // Invalidate weight-above cache for affected items (transitive supporters)
    let mut to_invalidate = HashSet::new();
    let mut queue: Vec<usize> = ctx.supported_by[new_idx].clone();
    while let Some(idx) = queue.pop() {
        if !to_invalidate.insert(idx) {
            continue;
        }
        for &below_idx in &ctx.supported_by[idx] {
            if !to_invalidate.contains(&below_idx) {
                queue.push(below_idx);
            }
        }
    }
    for idx in to_invalidate {
        ctx.weight_above_cache.remove(&idx);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_def(id: &str, w: f64, h: f64, d: f64, weight: f64) -> CargoItemDef {
        CargoItemDef {
            id: id.to_string(),
            name: id.to_string(),
            width_cm: w,
            height_cm: h,
            depth_cm: d,
            weight_kg: weight,
            color: "#FF0000".to_string(),
            blocks: None,
            max_stack_weight_kg: None,
            no_stack: None,
            no_flip: None,
        }
    }

    #[test]
    fn test_no_violations_without_constraints() {
        let def = make_def("box", 100.0, 50.0, 100.0, 10.0);
        let p1 = PlacedCargo {
            instance_id: 1,
            cargo_def_id: "box".to_string(),
            position_cm: Vec3::zero(),
            rotation_deg: Vec3::zero(),
        };
        let p2 = PlacedCargo {
            instance_id: 2,
            cargo_def_id: "box".to_string(),
            position_cm: Vec3::new(0.0, 50.0, 0.0),
            rotation_deg: Vec3::zero(),
        };

        let mut ctx = build_stack_context(&[p1], &[def.clone()]);
        let violations = check_stack_incremental(&mut ctx, &p2, &def);
        assert!(violations.is_empty());
    }

    #[test]
    fn test_stack_violation() {
        let mut def = make_def("box", 100.0, 50.0, 100.0, 20.0);
        def.max_stack_weight_kg = Some(10.0);

        let p1 = PlacedCargo {
            instance_id: 1,
            cargo_def_id: "box".to_string(),
            position_cm: Vec3::zero(),
            rotation_deg: Vec3::zero(),
        };
        let p2 = PlacedCargo {
            instance_id: 2,
            cargo_def_id: "box".to_string(),
            position_cm: Vec3::new(0.0, 50.0, 0.0),
            rotation_deg: Vec3::zero(),
        };

        let mut ctx = build_stack_context(&[p1], &[def.clone()]);
        let violations = check_stack_incremental(&mut ctx, &p2, &def);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].max_stack_weight_kg, 10.0);
        assert_eq!(violations[0].actual_stack_weight_kg, 20.0);
    }

    #[test]
    fn test_no_stack() {
        let mut def = make_def("fragile", 100.0, 50.0, 100.0, 5.0);
        def.no_stack = Some(true);

        let p1 = PlacedCargo {
            instance_id: 1,
            cargo_def_id: "fragile".to_string(),
            position_cm: Vec3::zero(),
            rotation_deg: Vec3::zero(),
        };
        let p2 = PlacedCargo {
            instance_id: 2,
            cargo_def_id: "fragile".to_string(),
            position_cm: Vec3::new(0.0, 50.0, 0.0),
            rotation_deg: Vec3::zero(),
        };

        let mut ctx = build_stack_context(&[p1], &[def.clone()]);
        let violations = check_stack_incremental(&mut ctx, &p2, &def);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].max_stack_weight_kg, 0.0);
    }
}
